### PR TITLE
Methods without #[payable] cannot accept assets

### LINKED ISSUE #742: Methods without `#[payable]` annotation cannot accept coins
Once https://github.com/FuelLabs/sway/pull/3450 is merged the SDK also needs to forbid accepting coins in non-payable functions. The current plan is to introduce function attributes into the JSON ABI schema to mark payable functions as such. I'm happy to take any comments/suggestions on this.

Spec issue: https://github.com/FuelLabs/fuel-specs/issues/446
Compiler change: https://github.com/FuelLabs/sway/pull/3450

### PR DESCRIPTION
**Pull request by MujkicA**
Closes #742

`.call_params()` now returns a `Result` which either resolves to the `ContractCallHandler` instance or an `InvalidCallParameters` error. The latter happens when the user tries to set call parameters with an amount > 0 on a contract method that isn't payable.

Initially, I wanted to avoid implementing it like this. One previous attempt was to have a generic param <T> on `ContractCall` that would define the `call_params` field as a payable or non-payable version of the current `CallParameters` struct. But this meant that for `MultiContractCallHandler` handler to work, we would need to define a trait for the `CallParameters`. The resulting interface would provide some kind of setters/getters for both the payable and non-payable params version. If non-payable call params have setters for the amount, they have to be fallible, which is just the current solution but more bloat.

Open to suggestions.

### CODE REVIEW

**Comment by mohammadfawaz** [comment]
`forc 0.33.0` was out yesterday if you'd like to update the version in this PR.

File: packages/fuels-types/src/lib.rs [hunk 1]
```diff
@@ -52,6 +52,16 @@ pub struct ABIFunction {
     pub attributes: Option<Vec<Attribute>>,
 }
 
+impl ABIFunction {
+    pub fn is_payable(&self) -> bool {
+        if let Some(attributes) = &self.attributes {
+            return attributes.iter().any(|attr| attr.name == "payable");
+        }
+
+        false
```
Comment (segfault-magnet) [inline packages/fuels-types/src/lib.rs:61]: another way of writing it. I've no strong preference either way. I just wanted to point out the ability to iterate through an `Option` and flatten the contained `Vec`.
```suggestion
        self.attributes
            .iter()
            .flatten()
            .any(|attr| attr.name == "payable")
```

**Review by segfault-magnet — DISMISSED** [review]
Clean, simple. 

The first weekend I find myself not doing anything, I think I'll try and see what a type-safe alternative looks like. 

My gut tells me it's going to be ugly because it's going to make the call handlers generic -- but there is a small chance that a clean solution might be found.

**Review by segfault-magnet — DISMISSED** [review]

**Review by hal3e — DISMISSED** [review]

**Comment by digorithm** [comment]
CI is broken and branch is out of sync!

**Comment by MujkicA** [comment]
> CI is broken and branch is out of sync!

The forc bugfix has recently been merged, we still need a new release for our test contracts to build properly.

https://github.com/FuelLabs/sway/issues/3741

**Comment by Salka1988** [comment]
> > CI is broken and branch is out of sync!
> 
> The forc bugfix has recently been merged, we still need a new release for our test contracts to build properly.
> 
> [FuelLabs/sway#3741](https://github.com/FuelLabs/sway/issues/3741)

Maybe you could use FORC PATCH BRANCH: "" FORC PATCH REVISION: ""  from ci.yml for testing until they issue a new release?

**Review by Salka1988 — DISMISSED** [review]

**Comment by MujkicA** [comment]
#790 needs to be merged first.

**Comment by digorithm** [comment]
Can we sync this with master? 👀 https://github.com/FuelLabs/fuels-rs/pull/790 is closed, so this should be unblocked.

**Pull request closed without merging**